### PR TITLE
`flash_attn_wrapper` implementation

### PR DESCRIPTION
**Pull request by nsakkine**
Pull-request here introduces 'flash-attention' capability to AMD-AIG-AIMA/torchtune. Concretely, it adds a `flash_attn_wrapper` implementation which uses

- `flash_attn_func` for mask-free and causal-mask (e.g. llama 3.2 vision instruct encoder and decoder self-attention)
- `flash_attn_varlen_func` for masks which are reducible to trivial masks (all queries attend to all keys) using a variable sequence length implementation (e.g. llama 3.2 vision instruct decoder cross-attention for selected datasets)

For the latter case, the wrapper raises an exception if it receives a non-trivial (invalid) mask. This exception is captured in a higher-level function and compute is forwarded to `scaled_dot_product_attention` instead.

Furthermore, `VisionTransformer` and `VisionCrossAttentionMask`classes implement additional zero-padding to improve the performance of ROCm/flash-attention implementation.

### CODE REVIEW

File: torchtune/modules/transforms/_transforms.py
Comment (amd-mghanimi) [inline torchtune/modules/transforms/_transforms.py:1]: this seems to be a feature that we have to put some condition around it to enable/disable with FAv3 if needed. "if flash_attn_available()"

File: torchtune/modules/vision_transformer.py
Comment (amd-mghanimi) [inline torchtune/modules/vision_transformer.py:1]: same here, this may need a condition around it to enable/disable with FAv3 if needed.
"if flash_attn_available()"

File: torchtune/modules/attention_utils.py [hunk 1]
```diff
@@ -226,6 +340,124 @@ def _attention_call(
...
+        def _attention_call(
+            q: torch.Tensor,
+            k: torch.Tensor,
+            v: torch.Tensor,
+            mask: Optional[_MaskType],
+            dropout_p: float,
+            is_causal: bool,
+            with_kv_cache: bool,
+        ) -> torch.Tensor:
+            if is_flash_attn_available() and not with_kv_cache:
```
Comment (amd-mghanimi) [inline torchtune/modules/attention_utils.py:415]: It seems this IF condition has a large try...except. Perhaps it is better to have another wrapper like attempt_to_use_flash_attention (it is just for readability nothing else)

**Review by amd-mghanimi — COMMENTED** [review]
It would be nice to add the tory example as a test here to show that the mask_check works:
tests/torchtune/modules/test_attention_utils.py 
perhaps we could add the tests in a different PR.

File: torchtune/modules/attention_utils.py (hunk 1 above)
Comment (nsakkine) [reply torchtune/modules/attention_utils.py:444]: Cleaned up this, please re-check

File: torchtune/modules/transforms/_transforms.py
Comment (nsakkine) [reply torchtune/modules/transforms/_transforms.py:1]: Done

**Comment by nsakkine** [comment]
Implemented a unit test for `_check`

File: torchtune/modules/transforms/_transforms.py [hunk 2]
```diff
@@ -171,7 +172,9 @@ def __call__(
             # to a single image, so text tokens attend to all the image's tokens.
             # The mask is text_seq_len x mask_image_size if defined, otherwise
             # it uses current text/image sequence lengths.
-            mask = torch.zeros(text_seq_len, image_seq_len, dtype=torch.bool)
```
[the remark this answers is not in the document]
Comment (nsakkine) [reply torchtune/modules/transforms/_transforms.py:174]: True, should be fixed now. Please re-check
